beacon_node/beacon_chain/src/graffiti_calculator.rs

```diff
@@ -0,0 +1,399 @@
+            Ok(Self::UserSpecified(Graffiti::from(bytes)))
+        } else if cli_args.is_present("private") {
+            // When 'private' flag is present, use a zero-initialized bytes array.
+            Ok(Self::UserSpecified(Graffiti::from(bytes)))
+        } else {
+            // Use the default lighthouse graffiti if no user-specified graffiti flags are present
+            Ok(Self::default())
+        }
+    }
+
+    pub fn graffiti(&self) -> Graffiti {
+        match self {
+            GraffitiOrigin::UserSpecified(graffiti) => *graffiti,
+            GraffitiOrigin::Calculated(graffiti) => *graffiti,
+        }
+    }
+}
+
+impl Default for GraffitiOrigin {
+    fn default() -> Self {
+        let version_bytes = lighthouse_version::VERSION.as_bytes();
+        let trimmed_len = std::cmp::min(version_bytes.len(), GRAFFITI_BYTES_LEN);
+        let mut bytes = [0u8; GRAFFITI_BYTES_LEN];
+        bytes[..trimmed_len].copy_from_slice(&version_bytes[..trimmed_len]);
+        Self::Calculated(Graffiti::from(bytes))
+    }
+}
+
+impl Debug for GraffitiOrigin {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        self.graffiti().fmt(f)
+    }
+}
+
+pub struct GraffitiCalculator<T: BeaconChainTypes> {
+    pub beacon_graffiti: GraffitiOrigin,
+    execution_layer: Option<ExecutionLayer<T::EthSpec>>,
+    pub epoch_duration: Duration,
+    log: Logger,
+}
+
+impl<T: BeaconChainTypes> GraffitiCalculator<T> {
+    pub fn new(
+        beacon_graffiti: GraffitiOrigin,
+        execution_layer: Option<ExecutionLayer<T::EthSpec>>,
+        epoch_duration: Duration,
+        log: Logger,
+    ) -> Self {
+        Self {
+            beacon_graffiti,
+            execution_layer,
+            epoch_duration,
+            log,
+        }
+    }
+
+    /// Returns the appropriate graffiti to use for block production, prioritizing
+    /// sources in the following order:
+    /// 1. Graffiti specified by the validator client.
+    /// 2. Graffiti specified by the user via beacon node CLI options.
+    /// 3. The EL & CL client version string, applicable when the EL supports version specification.
+    /// 4. The default lighthouse version string, used if the EL lacks version specification support.
+    pub async fn get_graffiti(&self, validator_graffiti: Option<Graffiti>) -> Graffiti {
+        if let Some(graffiti) = validator_graffiti {
+            return graffiti;
+        }
+
+        match self.beacon_graffiti {
+            GraffitiOrigin::UserSpecified(graffiti) => graffiti,
+            GraffitiOrigin::Calculated(default_graffiti) => {
+                let Some(execution_layer) = self.execution_layer.as_ref() else {
+                    // Return default graffiti if there is no execution layer. This
+                    // shouldn't occur if we're actually producing blocks.
+                    crit!(self.log, "No execution layer available for graffiti calculation during block production!");
```

Comment:
Also wondering if this should be a `crit`? I assume a missing execution layer will cause a lot of other errors so this probably just becomes noise. A `warn` I think would be okay